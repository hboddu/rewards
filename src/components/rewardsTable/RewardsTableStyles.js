import { makeStyles } from 'tss-react/mui';

export const TableContainerStyles = makeStyles({ "name": "RewardsTableContainer" })(theme => ({
    grid: {
        display: 'flex',
        justifyContent: 'center',
        paddingTop: 10,
    },
    tableContainer: {
        backgroundColor: '#282C34',
        overflowX: 'auto'
    }
}))

export const TableStyles = makeStyles({ "name": "RewardsTable" })(theme => ({
    tableHead: {
        '& .MuiTableCell-root': {
            textAlign: 'center',
            color: 'burlywood',
            fontSize: 20,
            [theme.breakpoints.down("md")]: {
                fontSize: 18,
            },
            [theme.breakpoints.down("sm")]: {
                fontSize: 15,
                padding: 3,
            }
        },
    },
    tableBody: {
        '& .MuiTableCell-root': {
            textAlign: 'center',
            color: '#fff',
            fontSize: 18,
            [theme.breakpoints.down("md")]: {
                fontSize: 16,
            },
            [theme.breakpoints.down("sm")]: {
                fontSize: 14,
                padding: 2,
            },
        },
    },
}))