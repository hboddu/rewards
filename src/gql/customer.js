import { gql } from "@apollo/client";

export const GET_AllCustomers = gql`
  query {
    allCustomers {
      id
      customer_name
      transactions
    }
  }
`;