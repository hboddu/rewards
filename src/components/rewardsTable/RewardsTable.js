import React from 'react';
import { Table, TableHead, TableRow, TableCell, TableBody } from '@mui/material';
import { TableStyles } from './RewardsTableStyles';

const RewardsTable = ({ rewardsTableData }) => {

    const { classes } = TableStyles();

    return (
        <Table size="medium" aria-label="a dense table">
            <TableHead className={classes.tableHead}>
                <TableRow>
                    <TableCell align="center" colSpan={2} data-testid="customers">
                        Customers
                    </TableCell>
                    <TableCell align="center" colSpan={4} data-testid="reward">
                        Reward Points
                    </TableCell>
                </TableRow>
                <TableRow>
                    <TableCell sx={{ color: 'burlywood' }} data-testid="id">ID</TableCell>
                    <TableCell data-testid="name">Name</TableCell>
                    <TableCell data-testid="month1">Month1</TableCell>
                    <TableCell data-testid="month2">Month2</TableCell>
                    <TableCell data-testid="month3">Month3</TableCell>
                    <TableCell data-testid="total">Total Points</TableCell>
                </TableRow>
            </TableHead>
            <TableBody className={classes.tableBody}>
                {rewardsTableData?.map((item, index) => {
                    return (
                        <TableRow key={index} data-testid="tabledata">
                            <TableCell>{index + 1}</TableCell>
                            <TableCell data-testid="customer_name">{item.customer_name}</TableCell>
                            <TableCell data-testid="rewardsMonth1">{item.rewardsMonth1}</TableCell>
                            <TableCell data-testid="rewardsMonth2">{item.rewardsMonth2}</TableCell>
                            <TableCell data-testid="rewardsMonth3">{item.rewardsMonth3}</TableCell>
                            <TableCell data-testid="totalPoints">{item.totalPoints}</TableCell>
                        </TableRow>
                    )
                })}
            </TableBody>
        </Table>
    );
}

export default RewardsTable;
