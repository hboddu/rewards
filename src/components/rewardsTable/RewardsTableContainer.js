import React, { useState, useEffect } from 'react';
import { Paper, TableContainer, Container, Grid } from '@mui/material';
import { TableContainerStyles } from './RewardsTableStyles';
import { useQuery } from '@apollo/client';
import { GET_AllCustomers } from "../../gql/customer";
import { getTotalMonthlyPoints } from './RewardsTableUtil';
import RewardsTable from './RewardsTable';

const RewardsTableContainer = () => {

    const { classes } = TableContainerStyles();
    const [rewardsTableData, setRewardsTableData] = useState([]);
    // fetch data asynchronously using graphQL useQuery hook
    const { error, data } = useQuery(GET_AllCustomers, { fetchPolicy: "no-cache" });

    useEffect(() => {
        if (data?.allCustomers) {
            processCustomerTxnData(data.allCustomers);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [error, data])
  
    // Process the customer transaction data and build a state variable with all the data needed to pass to the table
    const processCustomerTxnData = (customersRewardsData) => {
        customersRewardsData.forEach((customer) => {
            customer.rewardsMonth1 = getTotalMonthlyPoints(customer.transactions.Month1);
            customer.rewardsMonth2 = getTotalMonthlyPoints(customer.transactions.Month2);
            customer.rewardsMonth3 = getTotalMonthlyPoints(customer.transactions.Month3);
            customer.totalPoints =  customer.rewardsMonth1 +  customer.rewardsMonth2 +  customer.rewardsMonth3;
        })
        setRewardsTableData(customersRewardsData);
    }

    return (
        <Container maxWidth="xl">
            {!error ? (
                <Grid container className={classes.grid}>
                    <Grid item lg={8} md={12} sm={12} xs={12}>
                        <TableContainer component={Paper} className={classes.tableContainer}>
                            <RewardsTable rewardsTableData={rewardsTableData} />
                        </TableContainer>
                    </Grid>
                </Grid>

            ) : <p data-testid="error">{error.message}</p>}
        </Container>
    );
}

export default RewardsTableContainer;
