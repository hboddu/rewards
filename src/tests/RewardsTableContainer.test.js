import { render, screen, cleanup } from '@testing-library/react';
import RewardsTableContainer from '../components/rewardsTable/RewardsTableContainer';
import { MockedProvider } from "@apollo/client/testing";
import { GET_AllCustomers } from "../gql/customer";
import renderer from 'react-test-renderer';
import { GraphQLError } from 'graphql';
import { getTotalMonthlyPoints } from '../components/rewardsTable/RewardsTableUtil';

afterEach(() => {
    cleanup(); 
})

describe("TableContainer Component" ,() => {

    test('returns monthly data', () => {
        render(<MockedProvider><RewardsTableContainer /></MockedProvider>);

        const data = [
            {
                "date": "2022-09-02",
                "amount": 30
            },
            {
                "date": "2022-09-05",
                "amount": 70
            },
            {
                "date": "2022-09-08",
                "amount": 80
            },
            {
                "date": "2022-09-15",
                "amount": 65
            },
            {
                "date": "2022-09-17",
                "amount": 120
            },
            {
                "date": "2022-09-21",
                "amount": 160
            },
            {
                "date": "2022-09-26",
                "amount": 90
            }
        ]

        const result = getTotalMonthlyPoints(data);
        expect(result).toBe(365);
    });

    test('Table Container Snapshot', () => {
        const TableContainerComponent = renderer.create(<MockedProvider><RewardsTableContainer /></MockedProvider>);
        let tree = TableContainerComponent.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("should show error UI", async () => {
        const customerMock = {
          request: {
            query: GET_AllCustomers,
            fetchPolicy: "no-cache"
          },
          error: new Error("An error occurred")
        };
        render(
          <MockedProvider mocks={[customerMock]} addTypename={false}>
            <RewardsTableContainer />
          </MockedProvider>
        );
        expect(await screen.findByTestId("error")).toBeInTheDocument();
    });

    test("should show Graphql error", async () => {
        const customerMock = {
            request: {
                query: GET_AllCustomers,
                fetchPolicy: "no-cache"
            },
            errors: [new GraphQLError("Error!")],
        };
        render(
          <MockedProvider mocks={[customerMock]} addTypename={false}>
            <RewardsTableContainer />
          </MockedProvider>
        );
        expect(await screen.findByTestId("error")).toBeInTheDocument();
    });
    
});
