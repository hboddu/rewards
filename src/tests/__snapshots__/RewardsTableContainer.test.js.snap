// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableContainer Component Table Container Snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
>
  <div
    className="MuiGrid-root MuiGrid-container css-vlfcgw-MuiGrid-root-RewardsTableContainer-grid"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-8 css-1hjr3y7-MuiGrid-root"
    >
      <div
        className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-a4h9to-MuiPaper-root-MuiTableContainer-root-RewardsTableContainer-tableContainer"
      >
        <table
          aria-label="a dense table"
          className="MuiTable-root css-rqglhn-MuiTable-root"
          role={null}
        >
          <thead
            className="MuiTableHead-root css-1ws0f6i-MuiTableHead-root-RewardsTable-tableHead"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                colSpan={2}
                data-testid="customers"
                scope="col"
              >
                Customers
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                colSpan={4}
                data-testid="reward"
                scope="col"
              >
                Reward Points
              </th>
            </tr>
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-19lwqm2-MuiTableCell-root"
                data-testid="id"
                scope="col"
              >
                ID
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                data-testid="name"
                scope="col"
              >
                Name
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                data-testid="month1"
                scope="col"
              >
                Month1
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                data-testid="month2"
                scope="col"
              >
                Month2
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                data-testid="month3"
                scope="col"
              >
                Month3
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                data-testid="total"
                scope="col"
              >
                Total Points
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root css-1ge5af5-MuiTableBody-root-RewardsTable-tableBody"
            role={null}
          />
        </table>
      </div>
    </div>
  </div>
</div>
`;
